### PR TITLE
[BE]: Enable ruff TCH rules and autofixes for better imports

### PR DESCRIPTION
**Pull request by Skylion007**
Automated fixes to put imports that are only used in type hints into TYPE_CHECKING imports. This also enables the RUFF TCH rules which will automatically apply autofixes to move imports in and out of TYPE_CHECKING blocks as needed in the future, this will make the initial PyTorch import faster and will reduce cyclic dependencies.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @LucasLLC

### CODE REVIEW

**Comment by ezyang** [comment]
I see quite a lot of noqas. What's incomplete about this autofix?

**Comment by Skylion007** [comment]
Nothing is incomplete about it. I left he noqas because they a reference a variable that is never used because it's in a False literal if statement inside of __main__. The only true false positive is some weird logic in one of the files with some dynamically imported modules in the else clause of a type checking block.

I disabled it in torch onnx because beartype is used there and last I checked beartype couldn't read imports in TYPE_CHECKING blocks, that may have changed recently though.

File: torch/__init__.py [hunk 1]
```diff
@@ -287,7 +287,7 @@ def load_shared_libraries(library_path):
 # Appease the type checker; ordinarily this binding is inserted by the
 # torch._C module initialization code in C
 if TYPE_CHECKING:
-    from . import _C as _C
+    from . import _C as _C  # noqa: TCH004
```
Comment (Skylion007) [inline torch/__init__.py:290]: Also this noqa as I think this is to make the linter happy without affecting import ordering.

File: torch/_dynamo/decorators.py [hunk 2]
```diff
@@ -1,3 +1,4 @@
+# ruff: noqa: TCH004
```
Comment (Skylion007) [inline torch/_dynamo/decorators.py:1]: This file has some weird dynamic imports in the else statement of TYPE_CHECKING.

**Review by XuehaiPan — APPROVED** [review]
LGTM.

**Comment by Skylion007** [comment]
@ezyang Mind granting approval as a core maintainer?

File: torch/__init__.py (hunk 1 above)
Comment (XuehaiPan) [reply torch/__init__.py:290]: We can remove the `TYPE_CHECKING` guard around `from torch import _C as _C`. The `_C` submodule is always imported and assigned to `torch._C`.

You can access `torch._C` immediately after `from torch._C import *`

https://github.com/pytorch/pytorch/blob/fb53cd64973167a3f4161d5d48fb11a022bf43f0/torch/__init__.py#L283-L290

The `_C` variable is also being used in the `torch` namespace during importing.

https://github.com/pytorch/pytorch/blob/fb53cd64973167a3f4161d5d48fb11a022bf43f0/torch/__init__.py#L611-L634

Comment (XuehaiPan) [reply torch/__init__.py:290]: > You can access `torch._C` immediately after `from torch._C import *`

Here is a minimal example of this:

```console
$ tree foo    
foo
├── __init__.py
├── bar.py
└── baz.py

1 directory, 3 files
```

```python
# foo/__init__.py

from foo.bar import *

assert 'bar' in globals()
print('foo.bar can be accessed after `import foo`')


from .baz import *

assert 'baz' in globals()
print('foo.baz can be accessed after `import foo`')
```

```python
# foo/bar.py

print('foo.bar imported')
```

```python
# foo/baz.py

print('foo.baz imported')
```

```python
In [1]: import foo
foo.bar imported
foo.bar can be accessed after `import foo`
foo.baz imported
foo.baz can be accessed after `import foo`

In [2]: foo.bar
Out[2]: <module 'foo.bar' from '/.../foo/bar.py'>

In [3]: foo.baz
Out[3]: <module 'foo.baz' from '/.../foo/baz.py'>
```

Comment (Skylion007) [reply torch/__init__.py:290]: Yeah, I think this was specifically to make a linter happy, I agree that from a Python point of view it's fine. I'd like to point out that many linters do not even support '*' imports.

File: torch/masked/maskedtensor/_ops_refs.py [hunk 3]
```diff
@@ -5,7 +5,7 @@
 import torch
 
 if TYPE_CHECKING:
-    import torch._ops
+    import torch._ops  # noqa: TCH004
```
[the remark this answers is not in the document]
Comment (XuehaiPan) [reply torch/masked/maskedtensor/_ops_refs.py:8]: Use `from torch._ops import OpOverload` instead.

```suggestion
    from torch._ops import OpOverload
```

File: torch/nn/parallel/replicate.py [hunk 4]
```diff
@@ -8,7 +8,7 @@
 
 if TYPE_CHECKING:
     import torch.jit
-    import torch.jit._state
+    import torch.jit._state  # noqa: TCH004
```
[the remark this answers is not in the document]
Comment (XuehaiPan) [reply torch/nn/parallel/replicate.py:11]: We can import the used types directly in `if TYPE_CHECKING:`.

File: torch/__init__.py (hunk 1 above)
Comment (XuehaiPan) [reply torch/__init__.py:290]: I mean remove the `if TYPE_CHECKING` block and add an explicit `from torch._C import _C` at the top level.

`from torch._C import _C` is a side effect of `from torch._C import *` under `torch/__init__.py`. So this change would be safe and make the code more readable.

File: test/distributed/_spmd/test_data_parallel.py [hunk 5]
```diff
@@ -12,7 +12,7 @@
 from torch.distributed._tensor import Replicate
 from torch.nn.parallel import DistributedDataParallel as DDP
 from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
-from torch.testing._internal.common_utils import run_tests
+from torch.testing._internal.common_utils import run_tests  # noqa: TCH001
```
Comment (ezyang) [inline test/distributed/_spmd/test_data_parallel.py:15]: cc @fegin This is because the spmd tests are all disabled per https://github.com/Skylion007/pytorch/commit/2f04aab1404383db1e1d160abf8149a1aeb8cccd

File: test/distributed/checkpoint/test_traverse.py [hunk 6]
```diff
@@ -5,7 +5,7 @@
 import torch
 
 import torch.distributed.checkpoint._traverse as _traverse
-from torch.distributed.checkpoint.metadata import STATE_DICT_TYPE
+from torch.distributed.checkpoint.metadata import STATE_DICT_TYPE  # noqa: TCH001
```
Comment (ezyang) [inline test/distributed/checkpoint/test_traverse.py:8]: Why not on this one? Here, it seems like it is only used in type annotation context

Comment (Skylion007) [reply test/distributed/checkpoint/test_traverse.py:8]: Oh, I just noqa'd all of tests, I'll fix it.

**Review by ezyang — APPROVED** [review]
okey dokey

**Comment by malfet** [comment]
I would be a bit cautious ignoring that many failures...

**Review by malfet — APPROVED** [review]
